stdlib/public/runtime/Metadata.cpp

```diff
@@ -2149,64 +2119,76 @@ static MetadataAllocator &getResilientMetadataAllocator() {
 #endif
 
 ClassMetadata *
-swift::swift_relocateClassMetadata(ClassMetadata *self,
-                                   size_t templateSize,
-                                   size_t numImmediateMembers) {
-  // Force the initialization of the metadata layout.
-  (void) self->getDescription()->getMetadataBounds();
-
-  const ClassMetadata *superclass = self->Superclass;
-
-  size_t metadataSize;
-  if (superclass && superclass->isTypeMetadata()) {
-    metadataSize = (superclass->getClassSize() -
-                    superclass->getClassAddressPoint() +
-                    self->getClassAddressPoint() +
-                    numImmediateMembers * sizeof(void *));
-  } else {
-    metadataSize = (templateSize +
-                    numImmediateMembers * sizeof(void *));
-  }
-
-  if (templateSize < metadataSize) {
-    auto rawNewClass = (char*) malloc(metadataSize);
-    auto rawOldClass = (const char*) self;
-    rawOldClass -= self->getClassAddressPoint();
+swift::swift_relocateClassMetadata(ClassDescriptor *description,
+                                   ClassMetadata *pattern,
+                                   size_t patternSize) {
+  auto bounds = description->getMetadataBounds();
+  auto metadataSize = bounds.getTotalSizeInBytes();
 
-    memcpy(rawNewClass, rawOldClass, templateSize);
-    memset(rawNewClass + templateSize, 0,
-           metadataSize - templateSize);
+  if (patternSize < metadataSize) {
+    auto bytes = (char*) malloc(metadataSize);
 
-    rawNewClass += self->getClassAddressPoint();
-    auto *newClass = (ClassMetadata *) rawNewClass;
-    newClass->setClassSize(metadataSize);
+    auto fullPattern = (const char*) pattern;
+    fullPattern -= pattern->getClassAddressPoint();
+    memcpy(bytes, fullPattern, patternSize);
+    memset(bytes + patternSize, 0,
+           metadataSize - patternSize);
```

Comment:
If we're calling `swift_initClassMetadata` properly on a relocated class now, I'm not sure how necessary this copy is.  But I guess you were planning on reworking all this in a follow-up?

Reply:
Yes, I’m going to rework it. There’s actually very little useful data that gets copied here.